spec_util/meld_test.go

```diff
@@ -485,3 +505,85 @@ func TestMeldDataFormats(t *testing.T) {
 
 	}
 }
+
+func TestMeldData(t *testing.T) {
+	testCases := []struct {
+		name     string
+		left     *pb.Data
+		right    *pb.Data
+		expected *pb.Data
+	}{
+		{
+			name: "merging oneof with primitive - merge primitive into existing variant",
+			left: wrapOneOf(&pb.OneOf{
+				Options: map[string]*pb.Data{
+					"d616v2O0iE8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+					}),
+					"va5tP-fnZF8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}},
+					}),
+				},
+				PotentialConflict: true,
+			}),
+			right: wrapPrim(&pb.Primitive{
+				Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+				FormatKind: "datetime",
+				Formats: map[string]bool{"timestamp": true},
+			}),
+			expected: wrapOneOf(&pb.OneOf{
+				Options: map[string]*pb.Data{
+					"d616v2O0iE8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+					}),
+					"va5tP-fnZF8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}},
+					}),
+				},
+				PotentialConflict: true,
+			}),
+		},
+		{
+			name: "merging oneof with primitive - merge oneof into primitive",
+			left: wrapPrim(&pb.Primitive{
+				Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+				FormatKind: "datetime",
+				Formats: map[string]bool{"timestamp": true},
+			}),
+			right: wrapOneOf(&pb.OneOf{
+				Options: map[string]*pb.Data{
+					"d616v2O0iE8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+					}),
+					"va5tP-fnZF8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}},
+					}),
+				},
+				PotentialConflict: true,
+			}),
+			expected: wrapOneOf(&pb.OneOf{
+				Options: map[string]*pb.Data{
+					"d616v2O0iE8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Int64Value{Int64Value: &pb.Int64{}},
+					}),
+					"va5tP-fnZF8=": wrapPrim(&pb.Primitive{
+						Value: &pb.Primitive_Uint64Value{Uint64Value: &pb.Uint64{}},
+					}),
+				},
+				PotentialConflict: true,
+			}),
+		},
+	}
+
+	for _, tc := range testCases {
+		left := tc.left
+		right := tc.right
+		err := MeldData(left, right)
+		assert.NoError(t, err, "[%s] failed to meld", tc.name)
+
+		if diff := cmp.Diff(tc.expected, left, cmpWitnessOptions...); diff != "" {
+			t.Errorf("[%s] found diff:\n %v", tc.name, diff)
+		}
+
+	}
```

Comment:
I've seen meld tests, like the one below, that suggest that meld should be commutative. Why not test both directions here instead of having two copies of the test?

https://github.com/akitasoftware/akita-libs/blob/85dbbc9c5fe1164f9b6f1265ef21adae837ce452/spec_util/meld_test.go#L239-L270

Reply:
Good idea, fixed in 3089a2d.